File: src/core/geez.ts
/**
 * Shared Ge'ez / Fidäl engine — for the Ethiosemitic languages written in the Ethiopic SYLLABARY-abugida (Amharic
 * `am`, Tigrinya `ti`). Each codepoint is a whole CV syllable (the vowel is baked into the glyph), so the g2p is a
 * flat fidel→CV lookup rather than a Brahmic matra/virama engine. Two features are UNWRITTEN in the script and
 * handled here: GEMINATION (phonemic but unmarked — left single, folded vs the referee) and the 6th-order vowel
 * [ɨ] (sadis), which is EPENTHETIC — inserted to break illegal clusters — and so is deleted on the surface
 * wherever the surrounding consonants form a legal cluster. The per-language differences (which fidel a codepoint
 * maps to — e.g. Tigrinya keeps the pharyngeals ħ/ʕ that Amharic merged to h/ʔ) live entirely in each language's
 * fidel.tsv; the epenthesis phonotactics are shared Ethiosemitic.
 */
import { loadTsvMap } from "./loadTsv.ts";

const VOWEL = "əuiaeɨoɐæ";
const VOWELS = new Set([...VOWEL]);
/** A phoneme token is a VOWEL if its BASE code point is one (so a modifier-bearing token like 'aː' still counts). */
const isVowelTok = (t: string | undefined): boolean => t !== undefined && VOWELS.has([...t][0] ?? "");

/** Split an IPA string into PHONEME tokens: an affricate (X͡Y) + any trailing modifiers (ʼ ʷ ʰ ̥ ː) count as ONE
 *  consonant, so cluster counting isn't fooled by the multi-codepoint spellings (d͡ʒ is one C, not three). */
function toPhonemes(s: string): string[] {
    const a = [...s];
    const out: string[] = [];
    for (let i = 0; i < a.length; i++) {
        let t = a[i]!;
        if (a[i + 1] === "͡") { t += a[i + 1]! + (a[i + 2] ?? ""); i += 2; } // affricate base ͡ base
        while (a[i + 1] !== undefined && "ʼʷʰ̥ː".includes(a[i + 1]!)) t += a[++i]!; // trailing modifiers
        out.push(t);
    }
    return out;
}

const NASAL = new Set([..."mnɲŋ"]);
const FRICATIVE = new Set([..."szʃʒfhħ"]);
/** Is the consonant sequence c1·c2 an illegal Ethiosemitic cluster that an epenthetic ɨ must break? Keyed on each
 *  token's BASE code point (so labialized sʷ/mʷ classify by s/m). Nasal + a homorganic stop (nb, nd, nɡ) is LEGAL;
 *  a fricative + ɾ (sɾ) is LEGAL; only a STOP + ɾ and nasal + nasal break. */
function illegalCluster(c1: string, c2: string): boolean {
    const b1 = [...c1][0]!, b2 = [...c2][0]!;
    if (b2 === "ɾ" && !FRICATIVE.has(b1)) return true; // stop + ɾ (ɡɨɾ, bɨɾ); fricative + ɾ (sɾ) is fine
    if (NASAL.has(b1) && NASAL.has(b2)) return true; // nasal + nasal (nɨɲ, mɨn)
    return false;
}

/**
 * Delete the epenthetic 6th-order [ɨ] where the surrounding consonants form a LEGAL cluster; keep it where deleting
 * would create an illegal one. The sadis [ɨ] is inserted to break clusters, so on the surface it survives only
 * where needed: (a) KEPT word-initially (ɨɡɨɾ 'foot'); (b) KEPT if deleting it would leave a WORD-FINAL consonant
 * cluster of ≥3 — an illegal complex coda (አምስት→amɨst, since 'mst#' is illegal; but MEDIALLY the cluster
 * resyllabifies, so አምስተኛ→amstəɲa keeps NO ɨ); (c) KEPT where deleting would abut a truly-illegal 2-cluster — a
 * STOP + /ɾ/ (ɡɨɾ, bɨɾ; a fricative + ɾ like sɾ is legal) or a nasal + nasal (nɨɲ, mɨn). Processed RIGHT-TO-LEFT so
 * an earlier ɨ sees the clusters a later deletion already created.
 */
export function deleteEpenthetic(s: string): string {
    const p = toPhonemes(s);
    const isCons = (t: string | undefined): boolean => t !== undefined && t !== "" && !isVowelTok(t);
    for (let i = p.length - 1; i >= 0; i--) {
        if (p[i] !== "ɨ") continue;
        if (p.slice(0, i).every((c) => c === "" || !isVowelTok(c))) continue; // word-initial ɨ is kept
        const wordFinal = !p.slice(i + 1).some(isVowelTok); // no vowel follows → word-final cluster
        let left = 0;
        for (let j = i - 1; j >= 0 && !isVowelTok(p[j]); j--) if (p[j] !== "") left++;
        let right = 0;
        for (let j = i + 1; j < p.length && !isVowelTok(p[j]); j++) if (p[j] !== "") right++;
        if (wordFinal && left + right >= 3) continue; // deleting → illegal ≥3 complex coda → keep
        // The IMMEDIATE non-empty neighbours become adjacent when the ɨ is deleted; a cluster forms only if BOTH
        // are consonants (a vowel on either side means no cluster — nothing to break).
        const prev = [...p.slice(0, i)].reverse().find((c) => c !== "");
        const next = p.slice(i + 1).find((c) => c !== "");
        if (isCons(prev) && isCons(next) && illegalCluster(prev!, next!)) continue; // deleting → illegal 2-cluster → keep
        p[i] = "";
    }
    return p.join("");
}

/** Build a fidel→CV word phonemizer for a Ge'ez-script language: `import.meta.url` + the fidel TSV filename. The
 *  Ethiopic wordspace ፡ (and any whitespace) is a word boundary — each part is phonemized independently. */
export function makeGeezG2P(metaUrl: string, fidelFile: string): (word: string) => string {
    let fidel: Map<string, string> | undefined;
    const map = (): Map<string, string> => (fidel ??= loadTsvMap(metaUrl, fidelFile));
    const word = (w: string): string => {
        if (/[፡\s]/u.test(w)) return w.split(/[፡\s]+/u).filter(Boolean).map(word).join(" ");
        let out = "";
        for (const ch of w.normalize("NFC")) out += map().get(ch) ?? "";
        return deleteEpenthetic(out).normalize("NFC");
    };
    return word;
}

File: csharp/Vernacula.Phonemizer/Core/Geez.cs
/**
 * Shared Ge'ez / Fidäl engine for the Ethiosemitic languages (Amharic, Tigrinya): a flat fidel→CV
 * lookup, plus deletion of the epenthetic 6th-order [ɨ] wherever the surrounding cluster is legal.
 * Ported from src/core/geez.ts — see that file for the corpus evidence.
 */
using System.Text;

namespace Vernacula.Phonemizer.Core;

public static class Geez
{
    private const string VOWEL = "əuiaeɨoɐæ";
    private static readonly HashSet<string> VOWELS = new(Js.CodePoints(VOWEL));

    /** A token is a VOWEL if its BASE code point is one, so a modifier-bearing token like 'aː' still counts. */
    private static bool IsVowelTok(string? t) =>
        t is not null && t.Length > 0 && VOWELS.Contains(Js.CodePoints(t)[0]);

    /** Split an IPA string into PHONEME tokens: an affricate (X͡Y) plus any trailing modifiers is ONE token,
     *  so cluster counting is not fooled by the multi-code-point spellings. */
    private static List<string> ToPhonemes(string s)
    {
        var a = Js.CodePoints(s);
        var outp = new List<string>();
        for (var i = 0; i < a.Count; i++)
        {
            var t = a[i];
            if (i + 1 < a.Count && a[i + 1] == "͡") { t += a[i + 1] + (i + 2 < a.Count ? a[i + 2] : ""); i += 2; }
            while (i + 1 < a.Count && "ʼʷʰ̥ː".Contains(a[i + 1], StringComparison.Ordinal)) t += a[++i];
            outp.Add(t);
        }
        return outp;
    }

    private static readonly HashSet<string> NASAL = new(Js.CodePoints("mnɲŋ"));
    private static readonly HashSet<string> FRICATIVE = new(Js.CodePoints("szʃʒfhħ"));

    /** Is c1·c2 an illegal Ethiosemitic cluster that an epenthetic ɨ must break? Keyed on each token's BASE
     *  code point, so labialized sʷ/mʷ classify by s/m. */
    private static bool IllegalCluster(string c1, string c2)
    {
        var b1 = Js.CodePoints(c1)[0];
        var b2 = Js.CodePoints(c2)[0];
        if (b2 == "ɾ" && !FRICATIVE.Contains(b1)) return true;
        if (NASAL.Contains(b1) && NASAL.Contains(b2)) return true;
        return false;
    }

    /**
     * Delete the epenthetic 6th-order [ɨ] where the surrounding consonants form a LEGAL cluster; keep it where
     * deleting would create an illegal one. ⚠ RIGHT-TO-LEFT on purpose, so an earlier ɨ sees the clusters a
     * later deletion already created — reversing the loop changes the output.
     */
    public static string DeleteEpenthetic(string s)
    {
        var p = ToPhonemes(s);
        bool IsCons(string? t) => t is not null && t != "" && !IsVowelTok(t);
        for (var i = p.Count - 1; i >= 0; i--)
        {
            if (p[i] != "ɨ") continue;
            var anyVowelBefore = false;
            for (var j = 0; j < i; j++)
                if (p[j] != "" && IsVowelTok(p[j])) { anyVowelBefore = true; break; }
            if (!anyVowelBefore) continue;
            var wordFinal = true;
            for (var j = i + 1; j < p.Count; j++)
                if (IsVowelTok(p[j])) { wordFinal = false; break; }
            var left = 0;
            for (var j = i - 1; j >= 0 && !IsVowelTok(p[j]); j--) if (p[j] != "") left++;
            var right = 0;
            for (var j = i + 1; j < p.Count && !IsVowelTok(p[j]); j++) if (p[j] != "") right++;
            if (wordFinal && left + right >= 3) continue;
            string? prev = null;
            for (var j = i - 1; j >= 0; j--) if (p[j] != "") { prev = p[j]; break; }
            string? next = null;
            for (var j = i + 1; j < p.Count; j++) if (p[j] != "") { next = p[j]; break; }
            if (IsCons(prev) && IsCons(next) && IllegalCluster(prev!, next!)) continue;
            p[i] = "";
        }
        return string.Concat(p);
    }

    private static readonly JsRe WordspaceTest = JsRegex.Compile("[፡\\s]", "u");
    private static readonly JsRe WordspaceSplit = JsRegex.Compile("[፡\\s]+", "u");

    /** Build a fidel→CV word phonemizer for a Ge'ez-script language. `moduleDir` is relative to src/, standing
     *  in for the TS `import.meta.url` (see LoadTsv.cs). */
    public static Func<string, string> MakeGeezG2P(string moduleDir, string fidelFile)
    {
        Dictionary<string, string>? fidel = null;
        Dictionary<string, string> Map() => fidel ??= LoadTsv.LoadTsvMap(moduleDir, fidelFile);
        string Word(string w)
        {
            if (WordspaceTest.IsMatch(w))
                return string.Join(" ",
                    WordspaceSplit.Re.Split(w).Where(part => part.Length > 0).Select(Word));
            var outp = new StringBuilder();
            foreach (var ch in Js.CodePoints(w.Normalize(NormalizationForm.FormC)))
                outp.Append(Map().TryGetValue(ch, out var v) ? v : "");
            return DeleteEpenthetic(outp.ToString()).Normalize(NormalizationForm.FormC);
        }
        return Word;
    }
}
